File: solutions/lua/square-root/2/square-root.lua
local SquareRoot = {}

function SquareRoot.square_root(radicand)
    -- return radicand ^ 0.5
    function internal(xn)
      xn1 = 0.5 * (xn + radicand / xn)
      if xn1 * xn1 == radicand then
        return xn1
      else
        return internal(xn1)
      end
    end
    return internal(radicand)
end

return SquareRoot

File: solutions/elixir/square-root/3/lib/square_root.ex
defmodule SquareRoot do
@moduledoc """
Utility to calculate the square root.
"""

  @doc """
  Calculate the integer square root of a positive integer
  """
  @spec calculate(radicand :: pos_integer) :: pos_integer
  def calculate(radicand) do
    do_calculate(radicand, 1)
  end

  @doc false
  defp do_calculate(radicant, x) when radicant == x * x, do: x
  defp do_calculate(radicant, x) do
    x_n = 0.5 * (x + radicant / x)
    do_calculate(radicant, x_n)
  end
end
